src/main/java/org/zalando/nakadi/service/CursorOperationsService.java

```diff
@@ -0,0 +1,199 @@
+import org.zalando.nakadi.exceptions.runtime.MyNakadiRuntimeException1;
+import org.zalando.nakadi.repository.TopicRepository;
+import org.zalando.nakadi.service.timeline.TimelineService;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.CURSORS_WITH_DIFFERENT_PARTITION;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_OFFSET_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_TIMELINE_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.PARTITION_NOT_FOUND;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.TIMELINE_NOT_FOUND;
+
+@Service
+public class CursorOperationsService {
+    private final TimelineService timelineService;
+
+    @Autowired
+    public CursorOperationsService(final TimelineService timelineService) {
+        this.timelineService = timelineService;
+    }
+
+    public Long calculateDistance(final NakadiCursor initialCursor, final NakadiCursor finalCursor)
+        throws InvalidCursorOperation {
+        // Validate query
+        if (!initialCursor.getPartition().equals(finalCursor.getPartition())) {
+            throw new InvalidCursorOperation(CURSORS_WITH_DIFFERENT_PARTITION);
+        } else if (initialCursor.getTimeline().getOrder() > finalCursor.getTimeline().getOrder()) {
+            throw new InvalidCursorOperation(INVERTED_TIMELINE_ORDER);
+        }
+
+        if (initialCursor.getTimeline().getOrder() == finalCursor.getTimeline().getOrder()) {
+            return getDistanceSameTimeline(initialCursor, finalCursor);
+        } else {
+            return getDistanceDifferentTimelines(initialCursor, finalCursor);
+        }
+    }
+
+    private long getDistanceSameTimeline(final NakadiCursor initialCursor, final NakadiCursor finalCursor) {
+        final long distance = numberOfEventsBeforeCursor(finalCursor) - numberOfEventsBeforeCursor(initialCursor);
+        if (distance < 0) {
+            throw new InvalidCursorOperation(INVERTED_OFFSET_ORDER);
+        }
+        return distance;
+    }
+
+    private long getDistanceDifferentTimelines(final NakadiCursor initialCursor, final NakadiCursor finalCursor) {
+        long distance = 0;
+
+        // get distance from initialCursor to the end of the timeline
+        distance += totalEventsInTimeline(initialCursor) - numberOfEventsBeforeCursor(initialCursor);
+
+        // get all intermediary timelines sizes
+        final String partitionString = initialCursor.getPartition();
+        for (int order = initialCursor.getTimeline().getOrder() + 1;
+             order < finalCursor.getTimeline().getOrder();
+             order++) {
+            final Timeline timeline = getTimeline(initialCursor.getTopic(), order);
```

Comment:
@rcillo Just found during testing that it dies here (topic is used instead of event type)